### PR TITLE
implementation of dynamic igemm fp16 wrw kernel for gfx908

### PR DESCRIPTION
**Pull request by shaojiewang**
## Add dynamic asm igemm kernel for conv-wrw with fp16 data type on gfx908.
Implementation for multiple dynamic igemm asm kernels with fp16 data type for wrw pass on gfx908. Test is done under **MIOpenDriver**. Test script and perf num data can be found here.
### Test script
[run-test.sh](https://amdcloud-my.sharepoint.com/:u:/g/personal/shaowang_amd_com/Eb55jKVfrONGmyh4XTqrg1kBGDp6ml2Am_YuekplhLd4eg?e=qSre2C)

### Perf num
[wrw_perf_link.xslx](https://amdcloud-my.sharepoint.com/:x:/g/personal/shaowang_amd_com/Eb62DEM2a4NOn_nhPndjSPYBsk8hBYNgFPwSOUDyJVeyfA?e=WPKYzy)

### CODE REVIEW

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:676]: Does this xDLOPs test excluded from non-gfx908 testing runs? (Ditto others)

Comment (shaojiewang) [reply test/CMakeLists.txt]: Hi, Artem,
I do not fully know how to exclude a test from non-gfx908 testing.
Could you please check if they are now excluded from non-gfx908 testing?

Comment (atamazov) [reply test/CMakeLists.txt]: > I do not fully know how to exclude a test from non-gfx908 testing.

IIRC `cmake ... -DMIOPEN_TEST_GFX908=On ...` tells that we build the library & tests for running on gfx908. Please make sure that gfx908-specific tests DO NOT run when this cmake variable is NOT set.

Comment (atamazov) [reply test/CMakeLists.txt]: ...then please check to see if it really works as expected ;)

**Comment by atamazov** [comment]
How urgent is this? Still no see comments from designated reviewers (?_?)

**Comment by asroy** [comment]
@shaojiewang Please add link to perf number

**Comment by shaojiewang** [comment]
> @shaojiewang Please add link to perf number

[wrw_perf_link.xslx](https://amdcloud-my.sharepoint.com/:x:/g/personal/shaowang_amd_com/Eb62DEM2a4NOn_nhPndjSPYBsk8hBYNgFPwSOUDyJVeyfA?e=WPKYzy)

**Comment by shaojiewang** [comment]
@atamazov @asroy @fronteer Please review, thanks.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:260]: Any specific reason for commenting this out?

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:405]: Please remind: Why you use strings in `tunables`? Integer values are much faster to compare, and you can use enums.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:795]: Why GetSolution returns different workspace size?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:795]: @atamazov Do you mean the reason why fp32 datatype has different workspace size compared to fp16?
As we use atomic add to split `n*ho*wo`, but gfx908 do not have fp16 atomic add instruction. So for fp32, we directly do atomic add. And for fp16 datatype, we need to create a buffer to do atomic add in fp32, and then cast the buffer to fp16 datatype.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:260]: Because these 2 kernels with tile size 64x64x32 has lower efficiency than 64x64x16 kernels.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:795]: No. AFAICS under certain conditions GetWorkspaceSize and GetSolution may return different workspace sizes. Am I right?

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:260]: [Recommendation] Can you please add relevant comment?

Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:261]: OK, added.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:796]: `if(ctx.IsFp32())
       result.workspce_sz = 0;
 else if(ctx.IsFp16())
       result.workspce_sz = log2_gemm_k_global_splits > 0 ? GetWorkspaceSize(ctx) : 0;`

Yes, I do workspace calculation by this way in GetSolution.
So, is it better to move these lines into GetWorkspaceSize to keep GetWorkspaceSize and GetSolution have the same return value?

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:406]: Ok, enum is used.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: Changed, the 2 function return the same value.

**Comment by atamazov** [comment]
Converted to draft to avoid spoiling docker image cache on CI. Please change this back to "normal" PR only after ending of CI moratorium (see mail). Then first merge from develop. Push the `[Ready for review]` button only after that. Thanks.

**Comment by shaojiewang** [comment]
Branch develop is merged.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: Yes, thanks.

GetWorkspaceSize and GetSolution must return the same workspace size for all applicable problem configs. This is one of the basic guarantees that all solvers must provide.

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:692]: [Recommendation] This condition is not necessary anymore (see `ALLOW_NONXDLOPS` and how it is used).

**Review by atamazov — COMMENTED** [review]
No objections, one recommendation

**Comment by atamazov** [comment]
Other reviewers, please have a look.

**Comment by atamazov** [comment]
On hold for #680.

**Comment by shaojiewang** [comment]
@atamazov Latest develop branch was merged. All comments in PR680 are implemented here. And `CallImplicitGemmWrwDynamic` will be refactor by a separate PR.

**Review by atamazov — CHANGES_REQUESTED** [review]
Almost good.

**Comment by atamazov** [comment]
@shaojiewang 
> And `CallImplicitGemmWrwDynamic` will be refactored by a separate PR.

Which means that I will need to manually test performance twice ;)

**Comment by atamazov** [comment]
@shaojiewang If this PR is **_required_** for 4.3, please change labeling to `urgency_blocker`.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:529]: `gemm_k_per_block` is retrieved from the tunable list. Normally it will not be 0. 
You are right. `MIOPEN_THROW` should be added anyway.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:607]: j = log2(gemm_k_per_block). 
In our kernels, gemm_k_per_block={32, 16, 8, 4}, so j={5,4,3,2}.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:751]: Yes, Thanks.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:796]: Yes, Thanks.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:529]: MIOPEN_THROW is added at the caller of this function.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:734]: Add a MIOPEN_THORW.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:742]: ditto

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:749]: @atamazov ditto

**Comment by shaojiewang** [comment]
@atamazov OK, I'll fix it in this PR. You asked me to make it fixed in a separate PR in issue #756 .

**Comment by shaojiewang** [comment]
`CallImplicitGemmWrwDynamic` is refactored.

**Comment by atamazov** [comment]
> @atamazov OK, I'll fix it in this PR. You asked me to make it fixed in a separate PR in issue #756 .

Yes, sometimes I confuse both myself and others))

**Comment by shaojiewang** [comment]
> > @atamazov OK, I'll fix it in this PR. You asked me to make it fixed in a separate PR in issue #756 .
> 
> Yes, sometimes I confuse both myself and others))

@atamazov BTW, it is refactored.

**Comment by shaojiewang** [comment]
For CI test, I always get timeout error like, `ERROR: Error fetching remote repo 'origin'`.
@atamazov Could you please take a look at this problem?

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:585]: This dependence does not seem trivial. 

[General rule] If you can't make the dependency obvious (programmatically), then please add a comment at least.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:905]: [Recommendation] This leads to copying of the conv_problem instance by value into both Factory and Invoker lambda instances. What about this:
```C++
auto opShapeArgs = ComputeDynamicIGemmWrwKernelArgs(conv_problem, log2_gemm_k_global_splits);

if(conv_problem.IsFp32())
{
    result.invoker_factory = [=](const std::vector<Kernel>& kernels) { ... }
}
else if(conv_problem.IsFp16() && log2_gemm_k_global_splits > 0)
{
    result.invoker_factory = [=](const std::vector<Kernel>& kernels) { ... }
}
else // fp16 && log2_gemm_k_global_splits <= 0
{
    result.invoker_factory = [=](const std::vector<Kernel>& kernels) { ... }
}
```
You'll have 3 smaller & simpler Factories/Invokers instead of one big. And no things like:
```
// replace dw by workspace for fp16 datatype
opArgs[1] = workSpace;
```

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:887]: [Note] data_ctx is used in cases like
```
data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
```
Obtaining "data_something" from "WrWInvokeParams" looks like nonsense ;) Perhaps `invoke_params` is still the best choice...

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:919]: Please check workspace size when you use it, like this:
```C++
if(invoke_params.workSpaceSize < workspaceSize)
    MIOPEN_THROW("Not enough workspace for invoker of ConvAsmImplicitGemmGTCDynamicWrwXdlops");
```
This is a general rule for Invokers.

**Review by atamazov — CHANGES_REQUESTED** [review]
Almost done.

**Comment by atamazov** [comment]
> For CI test, I always get timeout error like, `ERROR: Error fetching remote repo 'origin'`.
> @atamazov Could you please take a look at this problem?

MI100-5 is our of order. https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/184 opened.

[Note] You can do that too -- looking at CI (especially using BlueOcean) it is easy to find out which node it constantly failing.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: Done, please check latest commit. @atamazov

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: Done, please check 1c191a0. @atamazov

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: @atamazov I try to use `DataInvokeParams`. But it didn't work. As `DataInvokeParams.tensors` is type `ConvDataTensors`, and its `w` is a `ConstData_t`, I can not write to weight for wrw. But wrw needs to update weights.
Otherwise, we setup an another PR to change `ConvDataTensors` struct?

**Comment by atamazov** [comment]
FYI I just terminated CI job for 1c191a0

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:587]: @atamazov Comment added.

**Comment by shaojiewang** [comment]
> FYI I just terminated CI job for [1c191a0](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/1c191a03bb03c888ea204208624f577a92fa5fc1)

Yes, thanks.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:850]: Please rename `workspce_sz` to `required_workspace_size` or something like that.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:892]: Please rename `data_ctx` to `invoke_params` or to `wrw_invoke_params` or something like that everywhere (no "data" in the name, please).

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:946]: Please create `workspaceDesc` outside of Factory. dwDesc is available there as `conv_problem.weights`.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:941]:
```suggestion
                if(workSpace == nullptr || workSpaceSize < required_workspce_size)
```

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:937]: Please remove this. I do not think that it is Invoker's job to check if Solver produces valid workSpaceSize.

**Review by atamazov — CHANGES_REQUESTED** [review]
Let's optimize a bit and make it cleaner.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: Good!

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: renamed.

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: renamed

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: moved

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp
Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp]: removed

**Comment by atamazov** [comment]
Unfortunately, this should wait until #858 is merged in.

**Comment by shaojiewang** [comment]
> Unfortunately, this should wait until #858 is merged in.

@atamazov As #858 is merged, besides merging develop branch, what should I do for this PR and #680 ?

**Comment by shaojiewang** [comment]
> Nothing wrt #680. Continue with this PR as usual. Thanks!

Recent comments are all addressed. @atamazov Could you please re-review?

**Review by atamazov — APPROVED** [review]
Code looks good. Now I need to perform some testing, because I do not see how this affects the performance of the library (as a whole).

**Comment by shaojiewang** [comment]
> Code looks good. Now I need to perform some testing, because I do not see how this affects the performance of the library (as a whole).

May I do it by my side and post the data in this PR?

**Comment by atamazov** [comment]
> > Code looks good. Now I need to perform some testing, because I do not see how this affects the performance of the library (as a whole).
> 
> May I do it by my side and post the data in this PR?

Of course, and that would meet my optimistic expectations))

**Comment by shaojiewang** [comment]
> > > Code looks good. Now I need to perform some testing, because I do not see how this affects the performance of the library (as a whole).
> > 
> > 
> > May I do it by my side and post the data in this PR?
> 
> Of course, and that would meet my optimistic expectations))

Should I use the shell script that you tested pr680, or should I just test some models and post the wall time and gpu time?

**Comment by atamazov** [comment]
I use that script with develop baseline and PR, collect CSVs and logs and then compare/analyze results (including diffing the logs). You can do the same, or you can do something else -- please judge yourself.

**Comment by atamazov** [comment]
## Performance testing of `dynamic_asm_igemm_wrw_fp16` commit bc3714641 vs. `develop` commit 00d57548c

Preconditions:
- MI100 node
- BUILD_DEV=off, release build
- MIOPEN_FIND_MODE=NORMAL
- MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_ALL_SOLUTIONS=1
- 93 FP16 configs.

Results:
- Find API, 1st run (filling the binary kernel cache)
  - FP16 WrW GPU time: +3.5% on average, max ~40%
  - FP16 Fwd/Bwd GPU time: on par.
  - FP16 Wall Time and Aux Wall time: on par.
- Find API, 2nd run (the binary kernel cache is full)
  - FP16 WrW GPU time: +3.0% on average, max ~60%
  - FP16 Fwd/Bwd GPU time: on par.
  - FP16 Wall Time and Aux Wall time: on par.
- Immediate mode API
  - FP16 WrW GPU time: +3.3% on average, max ~60%
  - FP16 WrW Wall Time/Aux Wall time: +10% on average
  - FP16 Fwd/Bwd GPU time: on par.
  - FP16 Wall Time and Aux Wall time: on par

Verdict: Ok.

**Pull request merged**
